### ISSUE TITLE
Custom function example

### ISSUE THREAD

**Issue by robdejonge** [jpmens/mqttwarn#404]
After having used it for quite a while now, I figured I would suggest a simple custom function to somehow be included into the default package. Would it be an idea to open up a folder where people can add their proposed custom functions in?

**What it does**
Let the user toggle notifications on a per-topic basis, using a simple web-based interface. New topics automatically get added. 

**Implementation** 
1. Every 'subscription' in `mqttwarn.ini` sets `filter = togglestate()` and a custom function is defined in the appropriate file in the `defaults` section. Code below.
2. A file `warntoggle.json` is used to store toggle state. I have this stored in the public web directory, and symlinked from the _mqttwarn_ directory. 
3. A simple Python script made available to change the settings. Code below. 

**Custom function**
```
import time
import copy
import json

def togglestate(topic, payload, section, srv): 
    filename = "warntoggle.json"
    default_topicblock = False

    try:
        with open(filename) as infile:    
            toggles = json.load(infile)
            infile.close()

        if topic in toggles: 
            # file found, topic found
            topicblock = toggles[topic]
	    srv.logging.debug('togglestate() was called from the ' + section + ' section and found ' + topic + ' in ' + filename)
        else:
            # file found, adding new topic
            toggles[topic] = default_topicblock
            with open(filename, 'w') as outfile: 
                json.dump(toggles, outfile)
                outfile.close()

            topicblock = default_topicblock
            srv.logging.debug('togglestate() was called from the ' + section + ' section, did not find ' + topic + ' in ' + filename
)
            srv.logging.debug('togglestate() added ' + topic + ' to ' + filename + ' with blocking set to ' + str(topicblock))

    except Exception as e: 
        # file not found or other error
        topicblock = default_topicblock 
        srv.logging.debug('togglestate() encountered an error: ' + str(e) )

    srv.logging.debug('togglestate() will return ' + str(topicblock))
    return topicblock
```

**Web page**
```
#!/usr/bin/python

import json

def index(req):

    filename = "/var/www/html/warntoggle/warntoggle.json"
    writetoggles = {}
    readtoggles = {} 
    scriptname = __file__.split('/')[-1:][0]

    # write form data to the file if applicable

    if bool(req.form): 

        for k,v in req.form.items():
            writetoggles[k] = eval(str(v))

        with open(filename, 'w') as outfile: 
            json.dump(writetoggles, outfile)
            outfile.close()

    # read data from the file

    with open(filename, 'r') as infile:
        readtoggles = json.load(infile)
        infile.close()

    # display the form

    html  = "<html><body><h1>mqttwarn notification toggles</h1>"
    html += "<table>"
    html += "<form action=\"" + scriptname + "\" method=\"post\">"

    for key in sorted(readtoggles.keys()):
        if readtoggles[key] == True:
            block_checked = " checked"
            notify_checked = ""
        else:
            block_checked = ""
            notify_checked = " checked"

        html += "<tr><td>" + key + "</td><td><input type='radio' name='" + key + "' value='True'" + block_checked + ">Block <input type='radio' name='" + key + "' value='False'" + notify_checked+ ">Notify</td></tr>"

    html += "</table><br/>"
    html += "<nobr><input type='submit' value='Submit'> <a href='" + scriptname + "'>reload</a></nobr>"
    html += "</form>"
    html += "</body></html>"

    return html 
```

**Comment by jpmens** [comment]
Those contributions are a good idea, thank you for that. I can think of two different possibitlities:

* either a `contrib/` directory into which people would PR their contribution in a subdirectory from there, or
* the wiki

Do people have a preference for either of these?

**Comment by robdejonge** [comment]
I think I’d prefer the Wiki, as it allows easier explanations. If I may :)

**Comment by jpmens** [comment]
Fair enough. Can you see if you can add a new page off [this one](https://github.com/jpmens/mqttwarn/wiki/Custom-functions)?

**Comment by robdejonge** [comment]
Sure. Evening here, will do tomorrow.

**Comment by amotl** [comment]
Dear Rob,

thanks a bunch for your contribution! While the Wiki might be a good place to put in some common descriptions and guidelines, the files itself should definitively go to the repository. The `examples/` folder has already been well established for that.

It would be good if you add the two files accompanied by an appropriate `.ini` file to another subdirectory like `examples/webnotify`, [1] and [2] are appropriate blueprints. We might think about adding respective `README.rst` files into the same folders instead of having the accompanied information "out-of-band" within the Wiki. So, we have everything in the same place and under revision control, which I consider to be a good thing.

Let me know if you need further assistance.

With kind regards,
Andreas.

[1] https://github.com/jpmens/mqttwarn/tree/master/examples/hiveeyes
[2] https://github.com/jpmens/mqttwarn/tree/master/examples/homie

**Comment by robdejonge** [comment]
Any chance somebody can post a default .ini file for me? I tried generating one with `make-config`, but doesn't seem to be working. I'll edit that file, create an example JSON file and post it all into a subdirectory to `examples` with a README file.

**Comment by amotl** [comment]
Dear Rob,

> I tried generating a default configuration file with `mqttwarn make-config`, but it doesn't seem to be working.

We probably should spit out a more bare-bone configuration file there. What exactly does not work on your end with it?

Maybe it's the `functions = 'samplefuncs.py'` setting? Or should we say `protocol = 4` these days?

With kind regards,
Andreas.

**Comment by robdejonge** [comment]
I'm not sure what does not work. Running the command doesn't seem to do anything. I tried again just now, it starts and just runs until I ctrl-c it half an hour later. No warnings, comments, or anything else. 

What I'd like to know is the absolute minimum .ini file would look like. I'll use that as the example .ini in the example folder I'll make for this.

**Comment by amotl** [comment]
Dear Rob,

I've tried to reproduce your issue.
```
$ mqttwarn make-config > mqttwarn.ini
```
Without any MQTT broker running, at least it shows something on STDERR:
```
$ mqttwarn
2020-04-01 14:06:32,888 INFO     [mqttwarn.commands        ] Starting mqttwarn
2020-04-01 14:06:32,888 INFO     [mqttwarn.commands        ] Log level is DEBUG
2020-04-01 14:06:32,890 INFO     [mqttwarn.core            ] Successfully loaded service "file"
2020-04-01 14:06:32,891 INFO     [mqttwarn.core            ] Successfully loaded service "log"
2020-04-01 14:06:32,892 DEBUG    [mqttwarn.core            ] Attempting connection to MQTT broker localhost:1883...
2020-04-01 14:06:32,892 DEBUG    [mqttwarn.core            ] Setting LWT to clients/mqttwarn...
2020-04-01 14:06:32,897 ERROR    [mqttwarn.core            ] Cannot connect to MQTT broker at localhost:1883: [Errno 61] Connection refused
```

We should investigate why any of those commands are blocking on your end until you cancel it by `CTRL+C`. However, I am a bit clueless right now.

With kind regards,
Andreas.

**Comment by jpmens** [comment]
@robdejonge are you sure no logging output is written?

**Comment by robdejonge** [comment]
With regards to the `mqttwarn make-config`: a lot of stuff is indeed showing up in the log file. I tried to figure out what was going on and why no file is being created, but was unable to conclude anything as it was happening in my live and configured situation. Perhaps it was not supposed to work in that setting anyway?

Anyway, on the original topic of this issue, I've created a pull request - https://github.com/jpmens/mqttwarn/pull/408. I'll close this issue and further discussion on the actual submission can be done on the pull request.
